broker/src/test/java/io/camunda/zeebe/broker/partitioning/PartitionJoinTest.java

```diff
@@ -0,0 +1,101 @@
+ */
+package io.camunda.zeebe.broker.partitioning;
+
+import static io.camunda.zeebe.broker.test.EmbeddedBrokerRule.assignSocketAddresses;
+
+import io.atomix.cluster.MemberId;
+import io.camunda.zeebe.broker.Broker;
+import io.camunda.zeebe.broker.SpringBrokerBridge;
+import io.camunda.zeebe.broker.system.SystemContext;
+import io.camunda.zeebe.broker.system.configuration.BrokerCfg;
+import io.camunda.zeebe.broker.test.TestActorSchedulerFactory;
+import io.camunda.zeebe.broker.test.TestClusterFactory;
+import java.nio.file.Path;
+import java.time.Duration;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Consumer;
+import org.assertj.core.api.Assertions;
+import org.awaitility.Awaitility;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.io.TempDir;
+
+final class PartitionJoinTest {
+  @Test
+  void canJoinPartition(@TempDir final Path tmp) {
+    // given
+    @SuppressWarnings("resource")
+    final var existingBroker =
+        buildBroker(
+            tmp.resolve("broker-0"),
+            brokerCfg -> {
+              final var clusterCfg = brokerCfg.getCluster();
+
+              clusterCfg.setClusterSize(1);
+              clusterCfg.setNodeId(0);
+              clusterCfg.setPartitionsCount(1);
+              clusterCfg.setReplicationFactor(1);
+            });
+
+    final var initialContactPoint =
+        existingBroker.getConfig().getNetwork().getInternalApi().getAdvertisedAddress();
+
+    @SuppressWarnings("resource")
+    final var joiningBroker =
+        buildBroker(
+            tmp.resolve("broker-1"),
+            brokerCfg -> {
+              final var clusterCfg = brokerCfg.getCluster();
+
+              // Joining broker knows about existing broker
+              clusterCfg.setInitialContactPoints(
+                  List.of(initialContactPoint.getHostName() + ":" + initialContactPoint.getPort()));
+
+              // Static configuration initially results in a broker without any partitions
+              clusterCfg.setClusterSize(1);
+              clusterCfg.setNodeId(1);
+              clusterCfg.setPartitionsCount(0);
+              clusterCfg.setReplicationFactor(0);
+            });
+    existingBroker.start().join();
+
+    // when
+    try (final var existing = existingBroker.start().join();
+        final var joining = joiningBroker.start().join()) {
+      final var partitionManager =
+          (PartitionManagerImpl) joining.getBrokerContext().getPartitionManager();
+
+      // then
+      Awaitility.await("Partition join completes")
+          .pollDelay(Duration.ofSeconds(1))
+          .pollInterval(Duration.ofMillis(500))
+          .untilAsserted(
+              () ->
+                  Assertions.assertThat(
+                          partitionManager.join(
+                              1, Map.of(MemberId.from("0"), 2, MemberId.from("1"), 1)))
+                      .succeedsWithin(Duration.ofSeconds(10)));
+    }
```

Comment:
Retrying here helps in cases where the joining member does not know about the existing member yet. Without retry, we would often run into "IllegalStateException: Cannot join cluster, because there are no other members in the cluster"

Reply:
I've adjusted this test to first wait for the membership service to report two known members before attempting the join. That seems to be more reliable.